File: src/core/gui/ToolBar.ts
import { Widget, WidgetOptions } from "./Widget";
import "./ToolBar.css";
import { LabelledControl } from "./LabelledControl";
import { div } from "./dom";

export class ToolBar extends Widget {
    readonly element = div({ className: "core_ToolBar" });
    readonly height = 33;
    readonly children: readonly Widget[];

    constructor(options?: WidgetOptions, ...children: Widget[]) {
        // noinspection SuspiciousTypeOfGuard
        super(options instanceof Widget ? undefined : options);

        this.element.style.height = `${this.height}px`;
        // noinspection SuspiciousTypeOfGuard
        this.children = options instanceof Widget ? [options, ...children] : children;

        for (const child of this.children) {
            this.disposable(child);

            if (child instanceof LabelledControl && child.label) {
                const group = div({ className: "core_ToolBar_group" });

                if (
                    child.preferred_label_position === "left" ||
                    child.preferred_label_position === "top"
                ) {
                    group.append(child.label.element, child.element);
                } else {
                    group.append(child.element, child.label.element);
                }

                this.element.append(group);
            } else {
                this.element.append(child.element);
            }
        }

        this.finalize_construction();
    }

    protected set_enabled(enabled: boolean): void {
        super.set_enabled(enabled);

        for (const child of this.children) {
            child.enabled.val = enabled;
        }
    }
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Toolbar.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.Node
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.dom.div

/**
 * Takes ownership of the given [children].
 */
class Toolbar(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    children: List<Widget>,
) : Widget(visible, enabled) {
    private val childWidgets = children

    override fun Node.createElement() =
        div {
            className = "pw-toolbar"

            childWidgets.forEach { child ->
                // Group labelled controls and their labels together.
                if (child is LabelledControl && child.label != null) {
                    div {
                        className = "pw-toolbar-group"

                        when (child.preferredLabelPosition) {
                            LabelPosition.Before -> {
                                addChild(child.label!!)
                                addChild(child)
                            }
                            LabelPosition.After -> {
                                addChild(child)
                                addChild(child.label!!)
                            }
                        }
                    }
                } else {
                    addChild(child)
                }
            }
        }

    companion object {
        init {
            @Suppress("CssUnusedSymbol", "CssUnresolvedCustomProperty")
            // language=css
            style("""
                .pw-toolbar {
                    box-sizing: border-box;
                    display: flex;
                    flex-direction: row;
                    align-items: center;
                    border-bottom: var(--pw-border);
                    padding: 3px 2px;
                }

                .pw-toolbar > * {
                    margin: 0 1px;
                }

                .pw-toolbar > .pw-toolbar-group {
                    margin: 0 3px;
                    display: flex;
                    flex-direction: row;
                    align-items: center;
                }

                .pw-toolbar > .pw-toolbar-group > * {
                    margin: 0 2px;
                }

                .pw-toolbar .pw-input {
                    height: 24px;
                }
            """.trimIndent())
        }
    }
}
